KitePrintSDK/src/main/java/ly/kite/analytics/Analytics.java

```diff
@@ -0,0 +1,659 @@
+
+        Country country = shippingAddress.getCountry();
+
+        if ( country != null )
+          {
+          jsonObject.put( JSON_PROPERTY_NAME_SHIPPING_COUNTRY,       nonNullString( country.displayName() ) );
+          jsonObject.put( JSON_PROPERTY_NAME_SHIPPING_COUNTRY_CODE2, nonNullString( country.iso2Code() ) );
+          jsonObject.put( JSON_PROPERTY_NAME_SHIPPING_COUNTRY_CODE3, nonNullString( country.iso3Code() ) );
+          }
+        else
+          {
+          jsonObject.put( JSON_PROPERTY_NAME_SHIPPING_COUNTRY,       "" );
+          jsonObject.put( JSON_PROPERTY_NAME_SHIPPING_COUNTRY_CODE2, "" );
+          jsonObject.put( JSON_PROPERTY_NAME_SHIPPING_COUNTRY_CODE3, "" );
+          }
+        }
+
+
+      ///// Cost /////
+
+      Set<String> supportedCurrencies = printOrder.getCurrenciesSupported();
+
+      if ( supportedCurrencies.contains( "GPB" ) )
+        {
+        jsonObject.put( JSON_PROPERTY_NAME_COST, printOrder.getCost( "GBP" ) );
+        }
+
+
+      // Job count
+
+      jsonObject.put( JSON_PROPERTY_NAME_JOB_COUNT, jobCount );
+      }
+    catch ( JSONException je )
+      {
+      Log.e( LOG_TAG, "Error setting JSON properties", je );
+      }
+    }
+
+
+  ////////// Constructor(s) //////////
+
+  private Analytics( Context context )
+    {
+    mContext = context;
+    }
+
+
+  ////////// Method(s) //////////
+
+  /*****************************************************
+   *
+   * Returns a unique id representing the user. This is
+   * generated and then persisted.
+   *
+   *****************************************************/
+  private String getUniqueUserId()
+    {
+    // If we don't have a cached id - see if we previously
+    // saved one and load it in. Otherwise generate a new
+    // one now and save it.
+
+    if ( mCachedUniqueUserId == null )
+      {
+      SharedPreferences sharedPreferences = mContext.getSharedPreferences( SHARED_PREFRENCES_NAME, Context.MODE_PRIVATE );
+
+      String uniqueUserId = sharedPreferences.getString( SHARED_PREFERENCES_KEY_UNIQUE_USER_ID, null );
+
+      if ( uniqueUserId == null )
+        {
+        uniqueUserId = UUID.randomUUID().toString();
+
+        sharedPreferences
+          .edit()
+            .putString( SHARED_PREFERENCES_KEY_UNIQUE_USER_ID, uniqueUserId )
+          .commit();
+        }
+
+      mCachedUniqueUserId = uniqueUserId;
+      }
+
+    return ( mCachedUniqueUserId );
+    }
+
+
+  /*****************************************************
+   *
+   * Returns a cached, or creates a new, JSON properties
+   * object.
+   *
+   * The same properties are uploaded with every analytics
+   * event, so it makes sense to cache them.
+   *
+   *****************************************************/
+  private JSONObject getPropertiesJSONObject()
+    {
+    // If we haven't already create the properties object - do
+    // so now.
+
+    if ( mCachedPropertiesJSONObject == null )
+      {
+      mCachedPropertiesJSONObject = new JSONObject();
```

Comment:
The caching of all properties in `mCachedPropertiesJSONObject` leads to properties "leaking" between different events i.e. on my SDK Launch event now has a bunch of properties that aren't applicable:

![screen shot 2015-07-20 at 11 40 04](https://cloud.githubusercontent.com/assets/760653/8774258/2c055e54-2ed4-11e5-8860-4a380a6e4950.png)

We definitely want to avoid this :)